autobump-maven/src/main/java/com/github/autobump/maven/model/MavenVersionRepository.java

```diff
@@ -9,45 +9,64 @@
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 
 @Getter
 public class MavenVersionRepository implements VersionRepository {
+    private static final transient int FILENOTFOUNDSTATUSCODE = 404;
     private final String baseUrl;
+    private final HttpClient httpClient;
 
     public MavenVersionRepository(String baseUrl) {
         this.baseUrl = baseUrl;
+        this.httpClient = HttpClient.newBuilder()
+                .connectTimeout(Duration.ofSeconds(2))
+                .followRedirects(HttpClient.Redirect.NEVER)
+                .version(HttpClient.Version.HTTP_2)
+                .build();
     }
 
     @Override
     public Set<Version> getAllAvailableVersions(Dependency dependency) {
-        try(BufferedReader in =
-                    new BufferedReader(
-                            new InputStreamReader(
-                                    getRepoUri(dependency)
-                                            .openConnection()
-                                            .getInputStream()
-                            )
-                    )
-        ) {
-            return new MetadataXpp3Reader()
-                    .read(in)
-                    .getVersioning()
-                    .getVersions()
-                    .stream()
-                    .map(MavenVersion::new)
-                    .collect(Collectors.toUnmodifiableSet());
-        } catch (XmlPullParserException e) {
-            throw new DependencyParserException("something went wrong while parseing the xml", e);
-        }catch (IOException e) {
-            return new HashSet<>();
+        try {
+            HttpResponse<InputStream> response = getInputStreamHttpResponse(dependency);
+            if (response.statusCode() == FILENOTFOUNDSTATUSCODE){
+                return new HashSet<>();
+            }
+            else {
+                return new MetadataXpp3Reader()
+                        .read(response.body())
+                        .getVersioning()
+                        .getVersions()
+                        .stream()
+                        .map(MavenVersion::new)
+                        .collect(Collectors.toUnmodifiableSet());
+            }
+        } catch (XmlPullParserException | IOException e) {
+            throw new DependencyParserException("something went wrong while parsing the xml", e);
+        }
+    }
+
+    private HttpResponse<InputStream> getInputStreamHttpResponse(Dependency dependency) {
+        try {
+            HttpRequest request = HttpRequest.newBuilder().GET().uri(getRepoUri(dependency).toURI())
+                    .build();
+            return httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+        } catch (URISyntaxException e) {
+            throw new WrongUrlException("Wrong URI Syntax", e);
+        } catch (InterruptedException | IOException e) {
```

Comment:
InterruptedException is always an annoying one. I also see there is a Sonar issue reported about this.

An example of where an InterruptedException might be thrown is for example:
1. The MavenVersionRepository sends a request to Maven Central to download the maven-metadata.xml file and waits for a response.
2. While waiting for the response, Autobump receives a SIGTERM signal to shutdown.
3. Java sends an interrupt to all the active threads including the thread that was waiting for the response.
4. Java interrupts threads by throwing an InterruptedException.
5. What you do in this case largely depends on the use case. 

In this case, I would not catch the exception in the getInputStreamHttpResponse method but at it to the method signature (throws InterruptedException).
I would catch the exception in the getAllAvailableVersions method, and in that catch I would just return an empty set (as if we found no versions at all), and re-interrupt the thread.

        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();

            return Collections.emptySet();
        }

Reply:
Yes, I was about to implement the interruption of the thread. Yet, we're considering another implementation now, with an async HttpRequest.